WIP FIX Dispatcher synchronization

rth opened the pull request:
Aims to address https://github.com/paris-saclay-cds/ramp-board/issues/154 

Fix by @glemaitre 

Not sure how to unit-test this. 

With this fix applied once a submission is sent there is a delay of 10-15 s for the server to respond. I'm not sure if it is related.

@glemaitre feel free to take over this PR..

pep8speaks commented:
Hello @rth! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`ramp-frontend/ramp_frontend/views/ramp.py`](https://github.com/paris-saclay-cds/ramp-board/blob/a9c22fad900fca2781c11836e96a95a576527696/ramp-frontend/ramp_frontend/views/ramp.py):

> [Line 431:80](https://github.com/paris-saclay-cds/ramp-board/blob/a9c22fad900fca2781c11836e96a95a576527696/ramp-frontend/ramp_frontend/views/ramp.py#L431): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (84 > 79 characters)

glemaitre commented:
Yep I think it was a quick hotfix but we should have something better.
I did not have time to commit it yet but basically we should copy the file within the function `add_submission`. Like this the copy will happen before the commit to the database and it will be safe.

Setting the state is a bit brittle. If you are not aware that you set the state of the submission, it could be tricky to know about it.

rth commented:
> I did not have time to commit it yet but basically we should copy the file within the function add_submission. Like this the copy will happen before the commit to the database and it will be safe

Can we not let the dispatcher create that file instead? That would remove the need for syncronization between the DB and the file system. Related to the discussion in https://github.com/paris-saclay-cds/ramp-board/issues/188

glemaitre commented:
Yep but we need to store the file into the database. This is a better long term solution I think.

glemaitre commented on ramp-frontend/ramp_frontend/views/ramp.py:
```diff
@@ -428,6 +428,8 @@ def sandbox(event_name):
                         dst=os.path.join(new_submission.path, filename)
                     )
+                # we delayed the state of the submission until that we copy the file
+                new_submission.set_state('new')
```
This function does not commit in the database. If you could either call `db.session.commit()` or use `ramp_database.tools.set_submission(db.session, new_submission.id, state)

glemaitre replied:
Could you do that on your server as well to check if it solves the issue?

rth commented:
Addressed your comment. 

Though there is an issue with this approach still. Turns out that the submissions that were marked with "check" permanently, were due to the fact that gunicorn worker timed-out (after 30 sec), particularly when the server had some load. Meaning that sending a submission can take more than 30 s which is a lot. 

Now I increased the gunicorn timeout to 1 hour, and this works, but understanding why it takes so long to make a submission with the frontend would be useful.

rth commented:
FWIW, it's a regular (physical) hard drive, there is no reason why interactions with the file system or writing into the DB would teake that long.

Maybe disabling wrting to the DB for user_interactions would help somewhat.
